libraries/AP_Proximity/AP_Proximity_LightWareSF40C.cpp

```diff
@@ -51,62 +57,87 @@ void AP_Proximity_LightWareSF40C::update(void)
     }
 
     // initialise sensor if necessary
-    bool initialised = initialise();
+    initialise();
 
     // process incoming messages
-    check_for_reply();
-
-    // request new data from sensor
-    if (initialised) {
-        request_new_data();
-    }
+    process_replies();
 
     // check for timeout and set health status
-    if ((_last_distance_received_ms == 0) || (AP_HAL::millis() - _last_distance_received_ms > PROXIMITY_SF40C_TIMEOUT_MS)) {
+    if ((last_distance_received_ms == 0) || ((AP_HAL::millis() - last_distance_received_ms) > PROXIMITY_SF40C_TIMEOUT_MS)) {
         set_status(AP_Proximity::Proximity_NoData);
     } else {
         set_status(AP_Proximity::Proximity_Good);
     }
 }
 
-// get maximum and minimum distances (in meters) of primary sensor
-float AP_Proximity_LightWareSF40C::distance_max() const
-{
-    return 100.0f;
-}
-float AP_Proximity_LightWareSF40C::distance_min() const
-{
-    return 0.20f;
-}
-
 // initialise sensor (returns true if sensor is succesfully initialised)
-bool AP_Proximity_LightWareSF40C::initialise()
+void AP_Proximity_LightWareSF40C::initialise()
 {
-    // set motor direction once per second
-    if (_motor_direction > 1) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_motor_direction();
-        }
+    // initialise sectors
+    if (!sector_initialised) {
+        init_sectors();
     }
-    // set forward direction once per second
-    if (_forward_direction != frontend.get_yaw_correction(state.instance)) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_forward_direction();
-        }
+
+    // exit immediately if we've sent initialisation requests in the last second
+    uint32_t now_ms = AP_HAL::millis();
+    if ((AP_HAL::millis() - last_request_ms) < 1000) {
+        return;
     }
-    // request motors turn on once per second
-    if (_motor_speed == 0) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_motor_speed(true);
-        }
-        return false;
+    last_request_ms = now_ms;
+
+    // if no replies in last 15 seconds reboot sensor
+    if ((now_ms > 30000) && (now_ms - last_reply_ms > 15000)) {
+        restart_sensor();
     }
-    // initialise sectors
-    if (!_sector_initialised) {
-        init_sectors();
-        return false;
+
+    // re-fetch motor state
+    request_motor_state();
+
+    // get token from sensor (required for reseting)
+    if (!got_token()) {
+        request_token();
+        return;
+    }
+
+    // if motor is starting up give more time to succeed or fail
+    if (sensor_state.motor_state < 3) {
```

Comment:
I think it's better to use enum values.

9.19. Motor state [106]
Reading this command will return the current state of the motor. This can be useful to debug or check start-up
conditions.
Value Description
1 Motor is preparing for start-up.
2 Motor is waiting for first 5 revolutions to
occur.
3 Motor is running normally.
4 Motor has failed to communicate

Reply:
sure, switched to use an enum, txs